docs/modules/ROOT/pages/deploy_related/integration_keystone.adoc

```diff
@@ -0,0 +1,98 @@
+
+When Keystone is enabled, the project_id attribute is not required in create requests
+because the project ID is derived from the authentication token."
+
+[#system-requirements]
+== Integration Requirements
+
+. Identify Neutron's algorithm that generates project id based on X-Auth-Token.
+. Make changes to API Gateway to support Authentication & Authorization with KeyStone, and support same way how project id is generated based on X-Auth-Token in API Gateway.
+. Integration testing with KeyStone.
+
+== Review of OpenStack Workflow
+https://docs.openstack.org/api-ref/identity/v3/index.html#check-token[Openstack keystone apis]
+
+In openstack neutron project, it use keystonemiddle to verify request token. .
+Verify request token have few steps.
+
+. get available keystone endpoints from keystone server
+. request local token from keystone server with keystone client conf in neutron.conf
+. verify request token from keystone server
+
+== Required Changes
+
+. Add a KeystoneAuthWebFilter class to fetch all api-gateway request token and verify token from keystone server and get project_id from verify api response.
+. Add keystone conf in application.properties file (https://docs.openstack.org/keystone/latest/user/supported_clients.html[keystone config example])
+
+== KeystoneAuthWebFilter Start Workflow
+image::keystone_filter_start_workflow.jpg["KeystoneAuthWebFilter Start Workflow", width=1024, link="keystone_filter_start_workflow.jpg"]
+
+== KeystoneAuthWebFilter Workflow
+image::keystone_filter_workflow.jpg["KeystoneAuthWebFilter Workflow", width=1024]
+
+== Token Cache Design
+Consider multi customer requests have a same token, it's not need to send a request to keystone to verfiy the same token for every request.
+A cache component can cache all verified token. Alcor will check cache first that the customer request token is verifed,
+it will return token info if cache have the customer token. If the customer token not in cache , it will send a request to keystone
+and check response and cache the token result.
+
+
+A cache data have a expired time to clear no use cache tokens.And Double the token expiration as the cache expiration time.
+This can prevent customer request from still carrying expired tokens for verification.
+
+== Data Schema
+How token store in cache： +
+A cache token key is token self. +
+A token entity include: +
+. string token: a token self +
+. date expireAt: the date which token expire at +
+. string user: the token user name  +
+. string userId: the token user id +
+. string domainId: the token user domain id +
+. string domainName: the token user domain name +
+. string projectName: the token project name +
+. string projectId: the token project id +
+. string projectDomain: the token project domain +
+. boolean invalid: the token whether valid from keystone, true： invalid  false: valid +
+. list<String> roles: the token user role names +
+
+examples: +
```

Comment:
We could make the example a bit pretty :-) by formatting it in an env.

Example: https://github.com/futurewei-cloud/alcor/blob/master/docs/modules/ROOT/pages/mgmt_services/private_ip_manager.adoc